core/src/backend/audio.rs

```diff
@@ -425,22 +441,72 @@ impl<'gc> AudioManager<'gc> {
             Avm2::dispatch_event(context, event, target.into());
         }
 
+        let domain = context.avm2.stage_domain();
+        let mut activation = Activation::from_domain(context.reborrow(), domain);
+        Self::update_generated_sounds(&mut activation);
+
         // Update sound transforms, if dirty.
         context.audio_manager.update_sound_transforms(context.audio);
     }
 
+    fn update_generated_sounds(activation: &mut Activation<'_, 'gc>) {
+        let mut states = Vec::new();
+        {
+            // Update the position of sounds, and remove any completed sounds.
+            for sound in &mut activation.context.audio_manager.sounds {
+                if let SoundInstanceSourceData::Generated(sound_state) = &mut sound.source_data {
+                    states.push(sound_state.clone());
+                }
+            }
+        }
+
+        for sound_state in &mut states {
+            let fps = activation.caller_movie_or_root().frame_rate().to_f32();
+
+            let mut ns = sound_state.next_samples.write().unwrap();
+            let mut pos = sound_state.position.write().unwrap();
+            // generating 3 frames worth of audio ahead to reduce hiccups
+            while ns.len() <= ((44100.0 * 2.0 * 3.0 / fps).ceil() as usize) {
+                let sample_data_evt =
+                    Avm2EventObject::sample_data_event(activation, "sampleData", *pos);
+                Avm2::dispatch_event(
+                    &mut activation.context,
+                    sample_data_evt,
+                    sound_state.sound_object.into(),
+                );
+                let data = sample_data_evt
+                    .get_public_property("data", activation)
+                    .unwrap()
+                    .coerce_to_object(activation)
+                    .unwrap();
+                let ba = data.as_bytearray().unwrap();
+
+                ba.set_position(0);
+                for _ in 0..ba.len() / 4 {
+                    ns.push_back(ba.read_float().expect("float can be read"));
```

Comment:
Shouldn't failing to read a float result in the read loop terminating early?

Reply:
that is to be tested :|